packages/compass-global-writes/src/store/reducer.ts

```diff
@@ -1,26 +1,251 @@
 import type { Action, Reducer } from 'redux';
+import type { GlobalWritesThunkAction } from '.';
+import { openToast } from '@mongodb-js/compass-components';
+import type { ManagedNamespace } from '../services/atlas-global-writes-service';
+
+export function isAction<A extends Action>(
+  action: Action,
+  type: A['type']
+): action is A {
+  return action.type === type;
+}
+
+export type CreateShardKeyData = Pick<
+  ManagedNamespace,
+  | 'customShardKey'
+  | 'isCustomShardKeyHashed'
+  | 'isShardKeyUnique'
+  | 'numInitialChunks'
+  | 'presplitHashedZones'
+>;
+
+enum GlobalWritesActionTypes {
+  SetIsManagedNamespace = 'global-writes/SetIsManagedNamespace',
+  ShardingInProgressStarted = 'global-writes/ShardingInProgressStarted',
+  ShardingInProgressFinished = 'global-writes/ShardingInProgressFinished',
+  ShardingInProgressErrored = 'global-writes/ShardingInProgressErrored',
+}
+
+type SetIsManagedNamespaceAction = {
+  type: GlobalWritesActionTypes.SetIsManagedNamespace;
+  isNamespaceManaged: boolean;
+};
+
+type ShardingInProgressStartedAction = {
+  type: GlobalWritesActionTypes.ShardingInProgressStarted;
+};
+
+type ShardingInProgressFinishedAction = {
+  type: GlobalWritesActionTypes.ShardingInProgressFinished;
+};
+
+type ShardingInProgressErroredAction = {
+  type: GlobalWritesActionTypes.ShardingInProgressErrored;
+};
 
 export enum ShardingStatuses {
   /**
    * Initial status, no information available yet.
    */
   NOT_READY = 'NOT_READY',
+
+  /**
+   * Namespace is not geo-sharded.
+   */
+  UNSHARDED = 'UNSHARDED',
+
+  /**
+   * Namespace is being sharded.
+   */
+  SHARDING = 'SHARDING',
 }
 
+export type ShardingStatus = keyof typeof ShardingStatuses;
+
 export type RootState = {
   namespace: string;
   isNamespaceSharded: boolean;
-  status: keyof typeof ShardingStatuses;
+  status: ShardingStatus;
+  createShardkey: {
+    isLoading: boolean;
+  };
```

Comment:
This was not part of the tech doc, but while working on this, I realzied we need to disable the create-shard-button and added this. as an alternative, we can also await on the dispatch action on the react side and while its waiting we disable the button.

Reply:
Can you clarify why is this not just part of sharding status? I'm maybe missing the purpose, but this seems like just part of the status transition for creation flow, so why are we moving this state into its own field?

Reply:
this is a state between `UNSHARDED` and `SHARDING`, time duration for which user clicks the create shard key and server acknowledges the change - keeps the button disabled, and for error shows the toast.

Reply:
So this loading state is mutually exclusive with all the other states, is that correct? If yes, in that case we should just add it as another state in list of available state transitions for the component, using multiple states to declare mutually exclusive, state-machine-like, states of the application can lead to potentially weird cases and hard to reason about logic as it basically allows application to be in multiple states at once.

Reply:
Yes that's correct. It does make sense to add these to the state as well, its just that I think the list will grow a little and in such case, i find it a bit hard to maintain the transitions.

But I will try to do that. Thanks for the suggestion.

Reply:
added in [23c1361](https://github.com/mongodb-js/compass/pull/6289/commits/23c136137c70829d9a7ca778edd358114de24a75)